From 86c094a3e4a9ea0f095d4199d79f231d967cd5fb Mon Sep 17 00:00:00 2001
From: Arie Brown <abrown@habana.ai>
Date: Tue, 30 Nov 2021 10:04:47 +0200
Subject: [PATCH] [SW-63523] gaudi2: fuse: print version

add support in print fuse version for gaudi2, similar to gaudi1.

Change-Id: I8ce4254c716673266038e5b75c368959ab7d664d
Signed-off-by: Arie Brown <abrown@habana.ai>
---

diff --git a/arch/gaudi2/common/asic.c b/arch/gaudi2/common/asic.c
index 97b9fc5..9fbe159 100644
--- a/arch/gaudi2/common/asic.c
+++ b/arch/gaudi2/common/asic.c
@@ -13,6 +13,7 @@
 #include <pcie_dbi_regs.h>
 #include <pcie_aux_masks.h>
 
+#include <string.h>
 #include <preboot.h>
 #include <pll.h>
 #include <uart.h>
@@ -43,6 +44,63 @@
 	regmap_writel(0x00000000, mmPCIE_AUX_DBI);
 }
 
+#define GAUDI2_FUSE_TABLE_SIZE	4
+#define GAUDI2_FUSE_DEVID_SIZE	8
+#define GAUDI2_FUSE_FAB_SIZE	2
+#define GAUDI2_FUSE_LOT_SIZE	6
+#define GAUDI2_FUSE_WAFER_SIZE	2
+#define GAUDI2_FUSE_XADDR_SIZE	2
+#define GAUDI2_FUSE_YADDR_SIZE	2
+
+struct hl_fuse_data {
+	union {
+		struct {
+			char table_ver[GAUDI2_FUSE_TABLE_SIZE];
+			char dev_id[GAUDI2_FUSE_DEVID_SIZE];
+			char hbm_config;
+			char fab[GAUDI2_FUSE_FAB_SIZE];
+			char process_type;
+			char lot[GAUDI2_FUSE_LOT_SIZE];
+			char wafer[GAUDI2_FUSE_WAFER_SIZE];
+			char xaddr[GAUDI2_FUSE_XADDR_SIZE];
+			char yaddr[GAUDI2_FUSE_YADDR_SIZE];
+		} __packed;
+		u32 data[7];
+	};
+};
+
+int hl_fuse_get_version(char *version, size_t size)
+{
+	struct hl_fuse_data fuse;
+	struct preboot_fuse_ver_data ver_data = {
+		.table_ver = fuse.table_ver,
+		.table_ver_sz = sizeof(fuse.table_ver),
+		.dev_id = fuse.dev_id,
+		.dev_id_sz = sizeof(fuse.dev_id),
+		.fab = fuse.fab,
+		.fab_sz = sizeof(fuse.fab),
+		.lot = fuse.lot,
+		.lot_sz = sizeof(fuse.lot),
+		.wafer = fuse.wafer,
+		.wafer_sz = sizeof(fuse.wafer),
+		.xaddr = fuse.xaddr,
+		.xaddr_sz = sizeof(fuse.xaddr),
+		.yaddr = fuse.yaddr,
+		.yaddr_sz = sizeof(fuse.yaddr),
+	};
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fuse.data); i++)
+		if (preboot_fuse_read(i, &fuse.data[i]))
+			return 1;
+
+	memcpy(hl_asic_sn, fuse.dev_id, sizeof(hl_asic_sn));
+
+	preboot_fuse_version_serialize(version, size, &ver_data, true);
+
+	return 0;
+}
+
 static struct hl_pll_cfg gaudi2_pll_cfg[] = {
 	{
 		.name = "PCI_PLL",
